thread_safe_function_counting_wrapped/node-addon-api/addon.cc

```diff
@@ -0,0 +1,158 @@
+    PopulateArg(env, std::get<I>(input), output[I]);
+    ArgsPopulator<TUPLE, I + 1u, N>::DoIt(env, input, output);
+  }
+  static void PopulateArg(Napi::Env env, int input, napi_value& output) { output = Napi::Number::New(env, input); }
+  static void PopulateArg(Napi::Env env, const std::string& input, napi_value& output) { output = Napi::String::New(env, input); }
+  // NOTE(dkorolev): Add more type wrappers or find the right way to do it within Napi.
+};
+
+template <class TUPLE, size_t N>
+struct ArgsPopulator<TUPLE, N, N> final {
+  static void DoIt(Napi::Env, const TUPLE&, std::vector<napi_value>&) {}
+};
+
+class NodeJSFunction final {
+ private:
+  struct Impl final {
+    // The `NodeJSContext` is captured into `std::shared_ptr<NodeJSFunction::Impl>`, to ensure proper cleanup order.
+    NodeJSContext context_;
+    Napi::ThreadSafeFunction function_;
+
+    Impl(NodeJSContext context, Napi::Function jsf)
+        : context_(context),
+          function_(Napi::ThreadSafeFunction::New(
+              context_.GetEnv(),
+              jsf,
+              "dkorolev_cpp_callaback",
+              0,  // Max queue size (0 = unlimited).
+              1,  // Initial thread count.
+              static_cast<void*>(nullptr),
+              [context](Napi::Env, void*, void*) {
+                // NOTE(dkorolev): The *IMPORTANT* part is to capture `context` by value here.
+                // If this is not done, the reference counter for the very `NodeJSContext` would drop to zero,
+                // the functions will get called, but the cleanup would fail, crashing the application.
+              },
+              reinterpret_cast<int*>(0))) {}
+    ~Impl() {
+      // NOTE(dkorolev): This `.Release()` would eventually call the finalization lambda, which, in its turn,
+      // would release the captured-by-copy `context`, ensuring the cleanup is happening as it should,
+      // first the captured functions, then by joining the async threads, and finally by setting the promise.
+      function_.Release();
+    }
+  };
+  std::shared_ptr<Impl> impl_;
+
+ public:
+  NodeJSFunction(NodeJSContext context, Napi::Function fun) : impl_(std::make_shared<Impl>(context, fun)) {}
+
+  template <typename... ARGS>
+  void operator()(ARGS&&... args) const {
+    auto args_as_tuple_to_copy = std::make_tuple(std::forward<ARGS>(args)...);
+    if (impl_->function_.BlockingCall(
+            reinterpret_cast<int*>(0), [args_as_tuple_to_copy](Napi::Env env, Napi::Function jsf, int*) {
+              std::vector<napi_value> params;
+              using tuple_t = decltype(args_as_tuple_to_copy);
+              params.resize(std::tuple_size<tuple_t>::value);
+              ArgsPopulator<tuple_t, 0, std::tuple_size<tuple_t>::value>::DoIt(env, args_as_tuple_to_copy, params);
+              jsf.Call(params);
+              // TODO(dkorolev): Process the return value as needed.
```

Comment:
Very neat implementation of "automagically" casting the C++ types to respective `Napi::Value`s.

You mention "process the return value". Currently the `operator()` returns `void`, so yes, the caller (eg. `f_odd(1,"one")` in your example) has no way of processing the return value of the JavaScript side.

The problem is that this is done asynchronously, so we can't return the immediate value. What about using an `std::promise<Napi::Value>` from the return value of `jsf.Call()`...? Just throwing some ideas out.

Reply:
Hi Kevin!

Thanks for chiming in! I've refactored the code since, you may find [this](https://github.com/C5T/Current/blob/stable/examples/nodejs/test.js) and [this](https://github.com/C5T/Current/blob/stable/examples/nodejs_external_server/run.js) more complete.

On a grander scheme of things, I'm looking into tighter C++ <=> V8 coupling these days, both in my day job and as a pet project, and would be glad to be of help if we could join our efforts. There's an open problem we have right now (can't create a JavaScript object with several wrapped C++ functions as the fields of this JSON object without these fuctions being garbage-collected), and I'm looking forward to finding the time to take a deeper look.

Thanks,
Dima